pyulog/db.py

```diff
@@ -0,0 +1,572 @@
+            cur = con.cursor()
+
+            cur.execute('''
+                SELECT COUNT(*)
+                FROM ULog
+                WHERE Id = ?
+                ''', (primary_key,))
+            count, = cur.fetchone()
+            cur.close()
+
+            return count == 1
+
+    # pylint: disable=super-init-not-called
+    def __init__(self, db_handle, primary_key=None, ulog=None, disable_str_exceptions=True):
+        '''
+        You always need a database handle, but there are two options for the
+        other parameters:
+        - For reading a log from the database, you just provide its primary
+          key. This will access the database and read the relevant info (if the
+          primary key exists).
+        - For storing a brand new log in the database, you provide its already
+          instantiated ULog object. Note that this does not immediately save
+          the object to the database: that is only done by explicitly calling
+          save().
+        You cannot supply both parameters.
+
+        Note that we purposefully don't call the superclass' constructor since
+        it would require a filename of a .ulg to load. Instead we use tests to
+        ensure compatibility. This is the reason why the seemingly useless
+        disable_str_exceptions option is also included.
+        '''
+
+        if primary_key is None and ulog is None:
+            raise ValueError('You must provide either a log primary key or a ULog object.')
+        if primary_key is not None and ulog is not None:
+            raise ValueError('You cannot provide both a log primary key and a ULog object.')
+
+        self._pk = primary_key
+        self._db = db_handle
+
+        attrs = [
+            ('_debug', False),
+            ('_file_corrupt', False),
+            ('_start_timestamp', 0),
+            ('_last_timestamp', 0),
+            ('_msg_info_dict', {}),
+            ('_msg_info_multiple_dict', {}),
+            ('_initial_parameters', {}),
+            ('_default_parameters', {}),
+            ('_changed_parameters', []),
+            ('_message_formats', {}),
+            ('_logged_messages', []),
+            ('_logged_messages_tagged', {}),
+            ('_dropouts', []),
+            ('_data_list', []),
+            ('_subscriptions', {}),
+            ('_filtered_message_ids', set()),
+            ('_missing_message_ids', set()),
+            ('_file_version', 0),
+            ('_compat_flags', [0] * 8),
+            ('_incompat_flags', [0] * 8),
+            ('_appended_offsets', []),
+            ('_has_sync', True),
+            ('_sync_seq_cnt', 0),
+        ]
+        ULog._disable_str_exceptions = disable_str_exceptions
+
+        if ulog is not None:
+            for attr_name, _ in attrs:
+                setattr(self, attr_name, getattr(ulog, attr_name))
+        else:
+            for attr_name, default_value in attrs:
+                setattr(self, attr_name, default_value)
+            self.load()
```

Comment:
To avoid this I'd rather make the `log_file` optional in the `ULog` initializer.